### PR TITLE
[Consensus Voting V2] Weighted signature aggregator interface

### PR DESCRIPTION
**Pull request by tarakby**
Implemented so far:
 - `WeightedSignatureAggregator` structure and first implementation: Signature aggregator aware of weights (but not of a threshold weight), aware of flow IDs, but agnostic of signature schemes. We would need two instance of this structure in the design: one for hotstuff, one for the beacon. 
 - `SignatureAggregator` structure and functions: BLS signature aggregator agnostic of weights and flow IDs, only responsible of pure cryptographic aggregations (used by hotstuff leaders) and verifying aggregates (used by followers to verify QC) 
 
 To be implemented in another PR:
 - implementation of `SignatureAggregator`. 
 - probable updates to `WeightedSignatureAggregator`

### CODE REVIEW

**Review by AlexHentschel — APPROVED** [review]
looks good. Some suggestions regarding API design

**Review by zhangchiqing — APPROVED** [review]

File: module/signature/aggregation.go [hunk 1]
```diff
@@ -10,14 +10,124 @@ import (
...
+//
+// Aggregation uses BLS scheme. Mitigation against rogue attacks is done using Proof Of Possession (PoP)
+// This module does not verify PoPs of input public keys, it assumes verification was done outside this module.
+//
+// Implementation of SignatureAggregator is not thread-safe, the caller should
+// make sure the calls are concurrent safe.
+type SignatureAggregatorSameMessage struct {
+	// TODO: initial incomplete fields that will evolve
+	message       []byte
+	hasher        hash.Hasher
```
[the remark this answers is not in the document]
Comment (tarakby) [reply module/signature/aggregation.go:27]: Yes, in theory we could store just the hash, or even better: just store the hash to point. Unfortunately there is no `Verify` api in the crypto lib that takes the hash, they all take the message as input. 
Avoiding hashing and mapping to point would save time, but that's still much less than what the rest of verification takes.

File: module/signature/aggregation.go [hunk 2]
```diff
@@ -10,14 +10,124 @@ import (
...
+//  - random error if the execution failed
+func (s *SignatureAggregatorSameMessage) VerifyAggregation(sig crypto.Signature, signers []int) (bool, error) {
+	panic("implement me")
+}
+
+// SetMessage sets a new message in the structure.
+// This allows re-using the same structure with the same keys for different messages.
+// Once called, the fuction empties all internally stored signatures.
+// The function is not thread-safe.
+func (s *SignatureAggregatorSameMessage) SetMessage(message []byte) { // we could also update the DST here
```
[the remark this answers is not in the document]
Comment (tarakby) [reply module/signature/aggregation.go:117]: This was meant for the use case of followers verifying the QC aggregated signature. 

In that use case, the list of possible signers during an epoch should be constant (if a node is slashed, its key is still on the list of possible signers, but the application logic should check it didn't sign). However the message changes at every block. In order to keep the same structure during the epoch (GC optim), and also optimize the public keys aggregation during an epoch (by comparing the delta of signers), setting a new message is needed, but setting a new list of keys is not.

File: module/signature/aggregation.go [hunk 3]
```diff
@@ -10,14 +10,124 @@ import (
...
+// VerifyAndAdd verifies the input signature under the stored message and stored
+// key at the input index. If the verification passes, the signature is added to the internal
+// signature state.
+// The function errors:
+//  - ErrInvalidInputs if the index input is invalid
+//  - ErrDuplicatedSigner if the signer has been already added
+//  - random error if the execution failed
+// The function does not return an error for any invalid signature.
+// The function is not thread-safe.
+func (s *SignatureAggregatorSameMessage) VerifyAndAdd(signer int, sig crypto.Signature) (bool, error) {
```
[the remark this answers is not in the document]
Comment (tarakby) [reply module/signature/aggregation.go:70]: No error is returned for an invalid signature. Errors are only returned for :
- duplicate signature
- invalid input which is a caller error (index is invalid for instance).
If there are no errors, the `bool` represents the validity of the signature. This is to be consistent with the verification API of crypto verifications in the repo.

[the remark this answers is not in the document]
Comment (tarakby) [reply module/signature/aggregation.go:70]: I updated the go doc to reflect this.

File: module/signature/aggregation.go [hunk 4]
```diff
@@ -10,14 +10,124 @@ import (
...
+//
+// The Aggregate function makes a sanity check on the aggregated signature and only
+// outputs valid signatures. This would detect if TrustedAdd has added any invalid
+// signature.
+// The function errors:
+//  - ErrInvalidInputs if the index input is invalid
+//  - ErrDuplicatedSigner if the signer has been already added
+//  - random error if the execution failed
+// The function is not thread-safe.
+func (s *SignatureAggregatorSameMessage) TrustedAdd(signer int, sig crypto.Signature) error {
```
[the remark this answers is not in the document]
Comment (tarakby) [reply module/signature/aggregation.go:84]: I don't think there is a bool output in `TrustedAdd`. There is one in `Verify` though (for the same reason as above). This is consistent with the crypto verification APIs (for instance a signature of a wrong length or serialization returns `false, nil` , it doesn't error)

File: module/signature/aggregation.go [hunk 5]
```diff
@@ -10,14 +10,124 @@ import (
...
+//
+// The module does not support signatures with multiplicity higher than 1. Each signer is allowed
+// to sign at most once.
+//
+// Aggregation uses BLS scheme. Mitigation against rogue attacks is done using Proof Of Possession (PoP)
+// This module does not verify PoPs of input public keys, it assumes verification was done outside this module.
+//
+// Implementation of SignatureAggregator is not thread-safe, the caller should
+// make sure the calls are concurrent safe.
+type SignatureAggregatorSameMessage struct {
```
[the remark this answers is not in the document]
Comment (tarakby) [reply module/signature/aggregation.go:24]: Yep, good idea. I think the sealing module needs this function.

File: module/signature/aggregation.go [hunk 6]
```diff
@@ -10,14 +10,124 @@ import (
...
+	// TODO: initial incomplete fields that will evolve
+	message       []byte
+	hasher        hash.Hasher
+	n             int                      // number of participants indexed from 0 to n-1
+	publicKeys    []crypto.PublicKey       // keys indexed from 0 to n-1, signer i is assigned to public key i
+	idToSignature map[int]crypto.Signature // signatures indexed by the signer index
```
[the remark this answers is not in the document]
Comment (tarakby) [reply module/signature/aggregation.go:30]: Yp that's right. The map just makes the look up faster than linear.

File: module/signature/aggregation.go [hunk 7]
```diff
@@ -10,14 +10,124 @@ import (
...
+}
```
[the remark this answers is not in the document]
Comment (tarakby) [reply module/signature/aggregation.go:31]: I guess that what `idToSignature map[int]crypto.Signature` also stores.  
The details of the implementation are coming in a another PR anyway.

File: consensus/hotstuff/signature/weighted_signature_aggregator.go [hunk 8]
```diff
@@ -1,39 +1,155 @@
...
-	panic("to be implemented")
+// TrustedAdd adds a signature to the internal set of signatures.
+//
+// It adds the signer's weight to the total collected weight and returns the total weight regardless
+// of the returned error.
+// The function errors if a signature from the signerID was already collected.
+// The function is thread-safe
+func (s *WeightedSignatureAggregator) TrustedAdd(signerID flow.Identifier, sig crypto.Signature) (uint64, error) {
+	// get the total weight safely
+	collectedWeight := s.TotalWeight()
```
[the remark this answers is not in the document]
Comment (tarakby) [reply consensus/hotstuff/signature/weighted_signature_aggregator.go:84]: You're right, I left it there by mistake. That was the original read before I moved it up to line 84.

File: module/signature/aggregation.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tarakby) [reply module/signature/aggregation.go:117]: I finally think `SetMessage` is confusing so I will remove it as you suggested, and think of a better way to address public keys aggregation.

File: consensus/hotstuff/signature/weighted_signature_aggregator.go [hunk 9]
```diff
@@ -1,39 +1,155 @@
...
+	// atomically update the signatures pool and the total weight
+	s.lock.Lock()
+	collectedWeight = s.totalWeight
+
+	// This is a sanity check because the upper layer should have already checked for double-voters.
+	_, ok = s.collectedIDs[signerID]
+	if ok {
+		s.lock.Unlock()
+		return collectedWeight, fmt.Errorf("SigneID %s was already added", signerID)
+	}
```
[the remark this answers is not in the document]
Comment (tarakby) [reply consensus/hotstuff/signature/weighted_signature_aggregator.go:106]: We don't really need to handle duplicates in the low level crypto, as we can assume the caller is responsible for duplicates (the function is `Trusted` anyway). The new signature could replace the previously added signature from the same signer. 
At the same time, the check is cheap and we could do it since the crypto component supports signatures of multiplicity one only.

File: consensus/hotstuff/signature/weighted_signature_aggregator.go [hunk 10]
```diff
@@ -1,39 +1,155 @@
...
+	totalWeight                               uint64                       // weight collected
+	lock                                      sync.RWMutex                 // lock for atomic updates
+	collectedIDs                              map[flow.Identifier]struct{} // map of collected IDs
 }
 
-var _ hotstuff.WeightedSignatureAggregator = &WeightedSignatureAggregator{}
+// NewWeightedSignatureAggregator returns a weighted aggregator initialized with the input data.
+//
+// A weighted aggregator is used for one aggregation only. A new instance should be used for each use.
+func NewWeightedSignatureAggregator(signerIDs []flow.Identifier,
```
[the remark this answers is not in the document]
Comment (tarakby) [reply consensus/hotstuff/signature/weighted_signature_aggregator.go:27]: thats' a good point. I updated the implementation.

File: consensus/hotstuff/signature/weighted_signature_aggregator.go [hunk 11]
```diff
@@ -1,39 +1,155 @@
...
-// WeightedSignatureAggregator aggregates the signatures
+// WeightedSignatureAggregator implements consensus/hotstuff.WeightedSignatureAggregator
 type WeightedSignatureAggregator struct {
+	*signature.SignatureAggregatorSameMessage                              // low level crypto aggregator, agnostic of weights and flow IDs
+	signerIDs                                 []flow.Identifier            // array of all signers IDs
+	idToIndex                                 map[flow.Identifier]int      // map node identifiers to indices
+	weights                                   map[flow.Identifier]uint64   // weight of each signer
+	totalWeight                               uint64                       // weight collected
+	lock                                      sync.RWMutex                 // lock for atomic updates
+	collectedIDs                              map[flow.Identifier]struct{} // map of collected IDs
```
[the remark this answers is not in the document]
Comment (tarakby) [reply consensus/hotstuff/signature/weighted_signature_aggregator.go:21]: Aggregating from a crypto perspective handles duplicates (signatures with multiplicity higher than 1). ` SignatureAggregatorSameMessage` is made to handle multiplicities equal to 1 only for the sake of simplicity. The constraint of checking duplicates comes from Hotstuff and double voting, so it makes sense IMO to have the check in a Hotstuff component, regardless of the crypto components underneath.

File: module/signature/errors.go [hunk 12]
```diff
@@ -7,4 +7,7 @@ import (
 var (
 	ErrInvalidFormat      = errors.New("invalid signature format")
 	ErrInsufficientShares = errors.New("insufficient threshold signature shares")
+
+	ErrInvalidInputs    = errors.New("invalid inputs")
```
[the remark this answers is not in the document]
Comment (tarakby) [reply module/signature/errors.go:11]: This is a good point, I totally agree we should not define duplicated structures, I didn't know about the existing error types. 
Reading the comment of the types, I think what I want to use is rather `engine.InvalidInputError`. The errors here are about misusing the APIs (for instance, a negative index) and not about invalid signatures/votes. Please note that an invalid signature (including wrong length or serialization) does not error, but outputs `false, nil`

[the remark this answers is not in the document]
Comment (tarakby) [reply module/signature/errors.go:11]: ( I re-used `engine.InvalidInputError` for now, but we can sync about this if needed )

File: consensus/hotstuff/signature/weighted_signature_aggregator.go [hunk 13]
```diff
@@ -1,39 +1,155 @@
...
+	n := len(signerIDs) // number of signers
+	if len(idToKey) != n || len(weights) != n {
+		return nil, fmt.Errorf("inconsistent inputs, got %d signers, %d keys, %d weights",
+			n, len(idToKey), len(weights))
+	}
+
+	// build a keys array
+	publicKeys := make([]crypto.PublicKey, 0, n)
+	for _, id := range signerIDs {
+		publicKeys = append(publicKeys, idToKey[id])
```
[the remark this answers is not in the document]
Comment (tarakby) [reply consensus/hotstuff/signature/weighted_signature_aggregator.go:44]: This is a tricky one. You're basically saying that 2 different nodes cannot share the same public key. I see 2 possible cases on how to handle this:
 - The protocol does not allow 2 nodes to share the same key: in this case the epoch transition/staking process should handle this and do not accept duplicate keys, and not a deep Hotstuff component. 
 - The protocol allows 2 nodes to share the same key: all components should not check if keys are duplicate.

File: module/signature/aggregation.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tarakby) [reply module/signature/aggregation.go:117]: `SetMessage` is not meant to be used in the Hotstuff leader case. In the leader case, I was assuming a new instance is used for every block. 
`SetMessage` was meant to be used by followers when checking QCs, but I'll delete the API as I think it's a bit confusing.

File: consensus/hotstuff/signature.go [hunk 14]
```diff
@@ -48,50 +48,49 @@ func (t SigType) Valid() bool {
...
 	Verify(signerID flow.Identifier, sig crypto.Signature) (bool, error)
 
-	// TrustedAdd adds an already verified signature, with weight for the given signer,
-	// and add it to the total weight, and returns the total weight that have been collected.
-	// return (1000, nil) means the signature has been added, and 1000 weight has been collected in total.
-	//   (1000 is just an example)
-	// return (1000, nil) means the signature is a duplication and 1000 weight has been collected in total.
+	// TrustedAdd adds a signature to the internal set of signatures.
+	// It adds the signer's weight to the total collected weight and returns the total weight regardless
+	// of the returned error.
```
Comment (AlexHentschel) [inline consensus/hotstuff/signature.go:62]: To me, the comment was a bit confusing depending on precedence of the `and` (which is not clearly defined in plain english). You could read it like 
* (It adds the signer's weight to the total collected weight and returns the total weight)
   
   regardless of the returned error.

```suggestion
	// TrustedAdd adds a signature to the internal set of signatures and adds the signer's weight to the total collected weight, iff the signature is _not_ a duplicate. 
	// The total weight of all collected signatures (excluding duplicates) is returned regardless
	// of any returned error.
```

### Suggestion:
Specify a sentinel error for duplicated signatures. Depending on the usage patter of the higher-level business logic, a duplicated signature could be a fatal bug or expected.

Comment (AlexHentschel) [reply consensus/hotstuff/signature.go:62]: BTW, in my mind 
1. signerID  corresponds to a known consensus participant but the signature is cryptographically invalid
2. signerID  does not correspond to a known consensus participant
3. duplicated signature

could _all_  be _expected_ during normal operations (depending on the usage patter of the higher-level business logic). 

I would suggest to follow the same API convention in all three cases, for all methods. Therefore, I would suggest to use the following convention:
* represent each case by a dedicated sentinel error 
* `TrustedAdd` can return any of the 3 sentinels 
* Change `Verify` to 
   ```golang 
   Verify(signerID flow.Identifier, sig crypto.Signature) error
   ```
    * it only returns nil if the signature is valid
    * case 1. and 2. are represented by sentinel errors
    * case 3 is irrelevant for this method 




I would suggest to update the API

**Comment by tarakby** [comment]
bors merge

**Comment by tarakby** [comment]
bors merge
